# Purpose: Analysis of cfDNA fragment coverage of patient data.
# Author: Yasamin Nouri Jelyani
# Concept from: Jonathan Broadbent
# Date: 2022-12-20
# Version: 0.1.0
# Bugs and Issues: Right now, we have low coverage as you can see
# in the daftaframe for coverage, we have 0 and 1 and max 5 coverage at a nucleotide position
# to get greater coverage, find the nucleotide position of TFBS. Aggregate all of those
# TFBS positions, and alighn then like the important first griffin paper image. site1, site2, site 3
# can be cromosome 8 bases 300-799, site 2 can be chromosome 22 at position 209-673,
# and site 3 can be chromosoem 3 at position 22-109. TF A binds to the mid position of
#all three sites. SO we alighn them like in the image, as if they are all
# in one location. Then, we mark that position as position 0, anything before as negative positions
# and anything after as positive positions. Then, we will have higher coverage, of around 300X since
# we have now more nucleotides at the position 0 because we put together all the
# TFBS at that position for TF A. Read Griffin for more details.


#------------------------Exported Function------------------------

#'
#'aggregate the coverage vectors across every window in the tfbs file
#'
#'
#' @param coverage a dictonary of chromosome names and coverage vectors
#' @param tfbs: path to the tfbs file
#' @param window: the size of the window to aggregate over
#'
#'
#' @return: aggregated coverage vector
#'
#'
#' @export
aggregate_coverage <- function(bed_file, tfbs, window){

  coverage <- get_coverage(bed_file)

  agg_coverage <- numeric(2*window + 1)

  if (! requireNamespace("readr", quietly=TRUE)) {
    install.packages("readr")
  }
  library(readr)
  #example:
  # TFBS_loc <- read_csv("inst/extdata/TFBS_loc.csv")
  TFBS_loc <- read_csv(tfbs)

  for (row in 1:nrow(tfbs)){
    start = tfbs$mid_position - window
    start = tfbs$mid_position + window
    agg_coverag = agg_coverage + coverage[[tfbs$chr]][start:end]
  }

  return(agg_coverage)

}





#' Get the coverage at every locus in the interval [start, end]
#'
#' Note 1: The data to generate examples are from the illumina website (4).
#'  This example data is not representative of real patient data due to
#'  limitations to access of real patient data.
#'  However, users can use their own real patient datasets for
#'  analysis.
#'  More examples can be found at https://github.com/Yasamin-Nourijelyani/CfDNAfragmentomics/tree/master/inst/extdata
#'
#' Note 2:
#' BED files can be generate from BAM files using bedtools bamtobed function
#' in the command line (3).
#'
#' Since BED files are commonly
#' generated by computational biologists and are easily convertible, from
#' BAM files, they can be used as input. Otherwise, tab seperated
#' .txt files that generate data in the same format of bed files
#' with the same columns as described in the parameters are also
#' excepted.
#'
#'
#' @param bed_file: The bed file to get the coverage from
#'  A dataframe for the a sample data
#' (unknown if cancerous or not individual)
#'  with at least 3 columns: first column is a string representing the chromosome
#'  number for the cfDNA for instance: "chr1",
#'  the second column are integer values indicating the start location of
#'  the read and third column are integer values indicating end location of
#'  the read.
#' These are the patient who we want to check if they have cancer or not.
#' The fragment lengths
#' of patients will be compared to these healthy data similar to what is done in
#' the paper by Katsman et. al (1).
#'
#'
#' #TODO:
#' add references to content above.
#'
#' @return A list of coverage values
#'
#' @references
#' 1- Katsman, E., Orlanski, S., Martignano, F., Fox-Fisher, I., Shemer,
#'  R., Dor, Y., Zick, A.,
#' Eden, A., Petrini, I., Conticello, S. G., & Berman, B. P. (2022).
#' Detecting cell-of-origin and cancer-specific methylation features of
#' cell-free DNA from Nanopore sequencing.
#' Genome biology, 23(1), 158.
#' https://doi.org.myaccess.library.utoronto.ca/10.1186/s13059-022-02710-1
#'
#' 2- “Mann Whitney U Test in R Programming.” GeeksforGeeks, 28 Dec. 2021,
#' https://www.geeksforgeeks.org/mann-whitney-u-test-in-r-programming/.
#'
#' 3- Quinlan, Aaron R., Kindlon, Neil. "bamtobed" readthedocs, May 26, 2022,
#'  https://bedtools.readthedocs.io/en/latest/content/tools/bamtobed.html
#'
#' 4- Illumina, https://support.illumina.com/downloads/nextera-flex-for-enrichment-BED-files.html
#'
#' 5- “The T-Test.” JMP,
#' https://www.jmp.com/en_ca/statistics-knowledge-portal/t-test.html#:~
#' :text=t%2DTest%20assumptions&amp;text=The%20data%20are%20continuous.,
#' The%20distribution%20is%20approximately%20normal.
#'
#'@examples
#' \dontrun{
#'
#'
#' sample_bed_file <- system.file("extdata", "d1.bed", package = "CfDNAfragmentomics")
#' sample_bed <- read.table(sample_bed_file, header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")
#'
#' ## basic usage
#'
#' # Example 1:
#'
#' ## Note: example does not work because the chromosome1 length is shoerter
#' than the chromosome length in sample bed file
#'
#' cov <- get_coverage(bed_file = sample_bed)
#' cov
#' # check for instance
#' cov[2985822, 1]
#'
#' }
#'
#'
#'
get_coverage <- function(bed_file){

  # create a dictionary with format: {chr1: 0...00, chr2: 0..00}
  coverage <- list()
  for (i in 1:22) {
    chrom <- paste0("chr", i)
    coverage[[chrom]] <- numeric(249*10**6)
  }

  #check if the input values are correct.
  if (! is.data.frame(bed_file)) {
    stop("Please put valid inputs for nucleosomeCoverage function")
  }


  #go along the indexes of the start_cov array
  for (row in 1:nrow(bed_file)){
    chr = bed_file[row, 1]
    start = bed_file[row, 2]
    end = bed_file[row, 3]
    for (i in range(start, end + 1)){
      coverage[[chr]][i] =  coverage[[chr]][i] + 1
    }

  }

  return(coverage)


}





# [END]

