# Purpose: Analysis of cfDNA fragment coverage of patient data.
# Author: Yasamin Nouri Jelyani
# Concept from: Jonathan Broadbent
# Date: 2022-12-20
# Version: 0.1.0
# Bugs and Issues: Right now, we have low coverage as you can see
# in the daftaframe for coverage, we have 0 and 1 and max 5 coverage at a nucleotide position
# to get greater coverage, find the nucleotide position of TFBS. Aggregate all of those
# TFBS positions, and alighn then like the important first griffin paper image. site1, site2, site 3
# can be cromosome 8 bases 300-799, site 2 can be chromosome 22 at position 209-673,
# and site 3 can be chromosoem 3 at position 22-109. TF A binds to the mid position of
#all three sites. SO we alighn them like in the image, as if they are all
# in one location. Then, we mark that position as position 0, anything before as negative positions
# and anything after as positive positions. Then, we will have higher coverage, of around 300X since
# we have now more nucleotides at the position 0 because we put together all the
# TFBS at that position for TF A. Read Griffin for more details.


#------------------------Exported Function------------------------

#'Return the tfbs from the csv inputted file
#'
#'@param csome_num the chromosome number of interest string like "chr1"
#'
#' @export

aggregate_coverage_over_tfbs <- function(TFBS_csv, sample_bed, offset=500){


  if (! requireNamespace("readr", quietly=TRUE)) {
    install.packages("readr")
  }
  library(readr)
  #example:
  # TFBS_loc <- read_csv("inst/extdata/TFBS_loc.csv")
  TFBS_loc <- read_csv(TFBS_csv)

  #unique and valid chromosomes
  csomes <- unique(TFBS_loc$chr)
  csomes <- csomes[! is.na(csomes)]
  #all coverage at a locus
  aggregate_data <- data.frame()


  for (chr in csomes){
    tfbs_for_csome <- TFBS_loc[TFBS_loc$chr == chr]
    #adjust midpoint with offset
    start = tfbs_for_csome$mid_position - offset
    end = tfbs_for_csome$mid_position + offset
    coverage <- get_coverage(sample_bed, chr, start, end)
    #bind row of coverage to dataframe
    rbind(aggregate_data, coverage)

  }
  #sum of all coverage
  final_high_coverage <- colSums(aggregate_data)



}





#' Nucleosome coverage plots for chromosome positions. Used to find Transcription
#' Factor Binding Sites (TFBS)
#'
#'TFBS can be used to subtype cancer DNA. Transcription Factors (TFs)bind to
#'specific regions of specific kinds of DNA, to allow for their transcription.
#'The binding of the TFs allow for generation of RNA from that DNA region.
#'It is important to note that regions that are binded by TFs are not bound
#'by nucleosomes (to allow for binding of the TF protein). However, other
#'regions of the DNA are bound by these nucleosomes, protecting the DNA from
#'degradation. cfDNA samples that do not have TF bound, do not have
#'nucleosomes bound at the region where TF usually binds. Hence, they are
#'not protected from nuclease degradation. This lead to the cfDNA to be cut at
#'the TF binding sites since those sites are unprotected. Hence, cfDNA fragment
#'ends if many of the cfDNA are cut at that certain region can be used as an
#'indicator for TF binding sites. Hence, coverage plots are used to detect the
#'amount of coverage by the cfDNA molecules on the DNA. If the coverage is low,
#'that could indicate the existance of a TFBS in that region.
#'The coverage plots are created using the nucleosomeCoverage data structure.
#'
#'
#' Note 1: The data to generate examples are from the illumina website (4).
#'  This example data is not representative of real patient data due to
#'  limitations to access of real patient data.
#'  However, users can use their own real patient datasets for
#'  analysis.
#'  More examples can be found at https://github.com/Yasamin-Nourijelyani/CfDNAfragmentomics/tree/master/inst/extdata
#'
#' Note 2:
#' BED files can be generate from BAM files using bedtools bamtobed function
#' in the command line (3).
#'
#' Since BED files are commonly
#' generated by computational biologists and are easily convertible, from
#' BAM files, they can be used as input. Otherwise, tab seperated
#' .txt files that generate data in the same format of bed files
#' with the same columns as described in the parameters are also
#' excepted.
#'
#'
#' @param sample_bed A dataframe for the a sample data
#' (unknown if cancerous or not individual)
#'  with at least 3 columns: first column is a string representing the chromosome
#'  number for the cfDNA for instance: "chr1",
#'  the second column are integer values indicating the start location of
#'  the read and third column are integer values indicating end location of
#'  the read.
#' These are the patient who we want to check if they have cancer or not.
#' The fragment lengths
#' of patients will be compared to these healthy data similar to what is done in
#' the paper by Katsman et. al (1).
#'
#' @param chr the chromosome of interest. Is a string of format for example
#' "chr1" for chromosome 1
#'
#' @param start the position of the cfDNA fragment
#'
#' @param end the ending position of the cfDNA fragment
#'
#' #TODO:
#' add references to content above.
#'
#' @return Return an dataframe of nucleosome coverage:
#' value refer to the number of cfDNA fragments that cover the region of the
#' nucleosome.
#'
#' @references
#' 1- Katsman, E., Orlanski, S., Martignano, F., Fox-Fisher, I., Shemer,
#'  R., Dor, Y., Zick, A.,
#' Eden, A., Petrini, I., Conticello, S. G., & Berman, B. P. (2022).
#' Detecting cell-of-origin and cancer-specific methylation features of
#' cell-free DNA from Nanopore sequencing.
#' Genome biology, 23(1), 158.
#' https://doi.org.myaccess.library.utoronto.ca/10.1186/s13059-022-02710-1
#'
#' 2- “Mann Whitney U Test in R Programming.” GeeksforGeeks, 28 Dec. 2021,
#' https://www.geeksforgeeks.org/mann-whitney-u-test-in-r-programming/.
#'
#' 3- Quinlan, Aaron R., Kindlon, Neil. "bamtobed" readthedocs, May 26, 2022,
#'  https://bedtools.readthedocs.io/en/latest/content/tools/bamtobed.html
#'
#' 4- Illumina, https://support.illumina.com/downloads/nextera-flex-for-enrichment-BED-files.html
#'
#' 5- “The T-Test.” JMP,
#' https://www.jmp.com/en_ca/statistics-knowledge-portal/t-test.html#:~
#' :text=t%2DTest%20assumptions&amp;text=The%20data%20are%20continuous.,
#' The%20distribution%20is%20approximately%20normal.
#'
#'@examples
#' \dontrun{
#'
#'
#' sample_bed_file <- system.file("extdata", "d1.bed", package = "CfDNAfragmentomics")
#' sample_bed <- read.table(sample_bed_file, header = FALSE, sep="\t",stringsAsFactors=FALSE, quote="")
#'
#' ## basic usage
#'
#' # Example 1:
#'
#' ## Note: example does not work because the chromosome1 length is shoerter
#' than the chromosome length in sample bed file
#'
#' cov <- nucleosomeCoverage(sample_bed = sample_bed, 237995947)
#' cov
#' # check for instance
#' cov[2985822, 1]
#'
#' }
#'
#'
#' @export
get_coverage <- function(sample_bed, chr, start, end){

  #check if the input values are correct.
  if (! is.data.frame(sample_bed) | ! is.numeric(chromosome_length)) {
    stop("Please put valid inputs for nucleosomeCoverage function")
  }

  # coverage array. Now, position 1 on this array is start
  chromosome_length = end - start
  coverage = c(1:chromosome_length) * 0

  #start locations of all cfDNA fragments
  start_cov <- sample_bed[ , 2][sample_bed[, 1] == chr]

  #adjust all indexes to fit on coverage vector
  start_cov <- start_cov - chromosome_length

  #end locations of all cfDNA fragment
  end_cov <- sample_bed[ , 3][sample_bed[, 1] == chr]

  #adjust all indexes to fit on coverage vector
  end_cov <- end_cov - chromosome_length

  #go along the indexes of the start_cov array
  for (i in seq_along(start_cov)){
    #if the coverage is a number and is a valid start and finish
    if(is.numeric(start_cov[i]) && is.numeric(end_cov[i]) && (end_cov[i] - start_cov[i] >= 0)){
      for (j in start_cov[i]:end_cov[i]){
        coverage[j] =  coverage[j] + 1

      }
    }

  }

  coverage <- as.data.frame(coverage)




  return(coverage)


}




# [END]

